StrongInject.Generator/Visitors/DependencyCheckerVisitor.cs

```diff
@@ -95,38 +94,36 @@ protected override bool ShouldVisitBeforeUpdateState(InstanceSource? source, Sta
-                // SingletonB depends on Func<NonSingletonA>
-                // Resolve NonSingletonA.
-                //
-                // Even though this should work fine, without clearing out non-singletons
-                // between each singleton we'd get an error.
-                //
-                // Clearing out non-singletons each time we visit a singleton is fine,
-                // since we'll still detect all cycles, just a bit later, but permits this one case.
-                state.CurrentlyVisitingNonSingletons = new();
-                
-                // Circular dependencies inside different scopes are not a problem since we assume any delegates are instantiated lazily.
-                // (Scope 0 represents the top level scope, and all other scopes represent dependencies resolved inside delegates.)
-                // However both checking dependencies and generating code for such cases is extremely complex.
-                // In the special case where a Singleton is included somewhere between the circular dependency,
-                // both are trivial ands require no special casing, so we allow only such cases.
-                if (state.InstanceSourcesScope.Depth != 0)
-                {
-                    state.CurrentlyVisitingSingletons = new();
-                }
+                // We don't care about circular dependencies between different scopes,
+                // since we assume the delegate is instantiated lazily.
+                // However we still need to track what we are currently visiting so we can warn if we
+                // can't generate correct code for the delegate.
+                // When the circular dependency crosses the boundary of a singleton this is always possible,
+                // so there's no need to track what we were visiting before the singleton.
+                state.CurrentlyVisiting = new();
             }
             base.UpdateState(source, ref state);
         }
 
         protected override bool ShouldVisitAfterUpdateState(InstanceSource source, State state)
         {
-            if (!state.AddCurrentlyVisiting(source))
+            if (!state.AddCurrentlyVisiting(source, state.InstanceSourcesScope, out var existingScope))
             {
+                if (source is DelegateSource delegateSource)
+                {
+                    var interveningDelegateParameters = state.InstanceSourcesScope.GetInterveningDelegateParameters(existingScope).ToList();
+                    if (interveningDelegateParameters.Count > 0)
+                    {
+                        _reportDiagnostic(WarnCircularDependencyWithInterveningDelegateParameter(_location, _target, delegateSource.DelegateType, interveningDelegateParameters));
+                    }
+                    return false;
+                }
+                if (state.InstanceSourcesScope.Depth != (existingScope.Depth))
+                {
+                    return true;
```

Comment:
`state.InstanceSourcesScope.Depth == existingScope.Depth` and ` _reportDiagnostic(CircularDependency(_location, _target, source.OfType))` feel more tightly connected to me than `return true;` does to the inverted condition.